### ISSUE TITLE
Generate Combined Edge Cases of Property's Parameters

### ISSUE THREAD

**Issue by jlink** [jlink/jqwik#97]
## Testing Problem

For a long time jqwik has been generating explicit edge cases for single arbitraries, e.g.
when generating integral values: MIN, MAX, -2, -1, 0, 1, 2. However, the combination of
edge cases of all parameters was not explicitly generated even though there is evidence
that those combinations will reveal bugs with a higher chance than purely random values.

## Suggested Solution

Make sure that all possible combinations of edge cases are generated for a property.

**Comment by jlink** [comment]
The problem of combinations/permutations of edge cases occurs on several levels:
- Generate all permutations of edge cases of a Property's parameters
- Use edge cases of embedded types (e.g. the value type of a list) to generate edge cases of a variable type, e.g. a singleton list with all edge cases of the concrete element type
- Combine edges cases in Combinators

Moreover, generated edge cases might appear before all randomly generated parameters or mixed-in with randomized parameters.

**Comment by jlink** [comment]
Example:

```
@Property
void propertyWithTwoParams(
    @ForAll List<Integer> intList,
    @ForAll @IntRange(min = -100, 0) anInt
) { ... }
```

The following combinatorials should be generated:

```
[], -100
[], -2
[], -1
[], 0
[-100], -100
[-100], -2
[-100], -1
[-100], 0
[-2], -100
[-2], -2
[-2], -1
[-2], 0
[-1], -100
[-1], -2
[-1], -1
[-1], 0
[[0], -100]
[[0], -2]
[[0], -1]
[[0], 0]
```

**Comment by jlink** [comment]
The feature is implemented and available in `1.3.0-SNAPSHOT`.

Running the following property

	@Property(tries = 100, edgeCases = EdgeCasesMode.FIRST)
	void propertyWithTwoParams(
		@ForAll List<Integer> intList,
		@ForAll @IntRange(min = -100, max = 0) int anInt
		) {
		String parameters = String.format("%s, %s", intList, anInt);
		System.out.println(parameters);
	}

will output:

```
[], -2
[], -1
[], 0
[], -100
[-2], -2
[-2], -1
[-2], 0
[-2], -100
[-1], -2
[-1], -1
[-1], 0
[-1], -100
[0], -2
[0], -1
[0], 0
[0], -100
[2], -2
[2], -1
[2], 0
[2], -100
[1], -2
[1], -1
[1], 0
[1], -100
[-2147483648], -2
[-2147483648], -1
[-2147483648], 0
[-2147483648], -100
[2147483647], -2
[2147483647], -1
[2147483647], 0
[2147483647], -100
[1589674531, -1280734788, -140838478, 1412584105, -1574796157, 93263621], -95
[0, 7, -24, 32, -35], -40
[], -68
[-676301465, 2016367551, 146545389, -11, 1001042157, 21805084], -47
[-45, -41, 44, 1254168564, 1203152559], -26
...
```

followed by more randomly generated parameters.

Edge cases are defined for most supported base types. Generic types (`List`, `Set`, `Map` etc.) will choose their edge cases based on the edge cases of underlying types.

The way how the full permutation of all parameters' edge cases are generated can be controlled by the `edgeCases` attribute of the `Property` annotation and can be one of:

- `MIXIN`: Edge cases are mixed into random parameters
- `FIRST`: All edge case permutations are tried before random generation
- `NONE`: No edge case permutations are tried

The default is `MIXIN` but it can be changed in `jqwik.properties`

Individual arbitraries will always mix-in edge cases from time to time.

Edge cases are also being combined and permutated when `Combinators` are used.
Also, most methods from `Arbitrary` - like `map()`, `filter()` and `flatMap()` - provide
sensible edge cases behaviour.
Thus, your self-made _domain-specific arbitraries get edge cases automatically_ by
combining edge cases from their underlying arbitraries!

Please experiment with the feature and give feedback.

**Comment by jlink** [comment]
Documented the current state in the user guide: https://jqwik.net/docs/snapshot/user-guide.html#generation-of-edge-cases
